metadata-ingestion/src/datahub/cli/generate_report/report_generator.py

```diff
@@ -0,0 +1,247 @@
+
+    def __init__(self, output_format: OutputFormatEnum, fileobj: TextIO):
+        self.output_format = output_format
+        self.fileobj = fileobj
+
+    def write(self, rows: Iterable[Dict]):
+        if self.output_format == OutputFormatEnum.CSV:
+            self._write_csv(rows)
+        else:
+            raise ValueError(f'unhandled output format: {self.output_format}')
+
+    def _write_csv(self, rows: Iterable[Dict]) -> None:
+        fieldnames = ['dataset', 'field', 'type', 'privacy_law']
+        writer = csv.DictWriter(self.fileobj, fieldnames=fieldnames)
+        writer.writeheader()
+        for row in rows:
+            # TODO: make this more generic rather than just privacy
+            # transform lists to be csv friendly
+            row['type'] = ', '.join(sorted(row['type']))
+            row['privacy_law'] = ', '.join(sorted(row['privacy_law']))
+            writer.writerow(row)
+
+class OutputExporter:
+    """Use this as a context manager. Write using `fileobj` property and call `export()`.
+    The `fileobj` will be closed on context manager exit.
+
+    >>> with OutputExporter(destination) as exporter:
+    >>>     exporter.fileobj.write(...)
+    >>>     exporter.export()
+    """
+
+    output_destination: OutputDestination
+    _fileobj: TextIO
+
+    def __init__(self, output_destination: OutputDestination):
+        self.output_destination = output_destination
+        destination_type = type(self.output_destination)
+        if destination_type == FileOutput:
+            self._fileobj = open(output_destination.filepath, 'w')
+        elif destination_type == S3Output:
+            self._fileobj = tempfile.NamedTemporaryFile('w')
+        else:
+            raise ValueError(f'unhandled output destination type: {destination_type}')
+
+    def __enter__(self) -> None:
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self._fileobj.close()
+
+    @property
+    def fileobj(self) -> TextIO:
+        return self._fileobj
+
+    def export(self):
+        if type(self.output_destination) == FileOutput:
+            # File should be written to, nothing to export
+            logger.info(f'Output at: {os.path.abspath(self._fileobj.name)}')
+        elif type(self.output_destination) == S3Output:
+            # set file to beginning before upload
+            # or else it will upload an empty file.
+            self.fileobj.seek(0)
+            bucket = self.output_destination.s3_bucket
+            key = datetime.datetime.utcnow().strftime(self.output_destination.s3_key)
+            S3FileUploader().upload(self.fileobj.name, bucket, key)
+        else:
+            raise ValueError(f'unhandled output destination: {self.output_destination}')
+
+class S3FileUploader:
+    s3_client: Any
+
+    def __init__(self, s3_client=None) -> None:
+        self.s3_client = s3_client or boto3.Session().client('s3')
+
+    def upload(self, src_filepath, s3_bucket, s3_key) -> None:
```

Comment:
Could we also upload to a static file path?

Reply:
yes because implementation wise it just runs `strftime` so if it has format strings, it will format, if not, nothing changes so it still works.

Reply:
done, wrote doc https://github.com/Affirm/datahub/pull/3/commits/3deac44ed9124d35c1364a60d1e1a94ea75de193#diff-bb3382b14deac0a215a7f92eb1b08fdc9c0c009ae1a7b5c70340332a16eca362R17-R18